File: csharp/910.smallest-range-ii.cs
/*
 * @lc app=leetcode id=910 lang=csharp
 *
 * [910] Smallest Range II
 */

using System;

// @lc code=start
public class Solution {
    public int SmallestRangeII(int[] A, int K) {
        int ret, left, right, n = A.Length;
        Array.Sort(A);
        left = A[0] + K;
        right = A[n - 1] - K;
        ret = A[n - 1] - A[0];
        for (int i = 0; i < n - 1; ++i) {
            int r = Math.Max(right, A[i] + K), l = Math.Min(left, A[i + 1] - K);
            ret = Math.Min(ret, r - l);
        }
        return ret;
    }
}
// @lc code=end

File: java/910.smallest-range-ii.java
/*
 * @lc app=leetcode id=910 lang=java
 *
 * [910] Smallest Range II
 */

// @lc code=start
class Solution {
    public int smallestRangeII(int[] A, int K) {
        int max, n = A.length, ret;
        Arrays.sort(A);
        max = A[n - 1];
        ret = A[n - 1] - A[0];
        for (int i = 1; i < n; ++i) {
            max = Math.max(max, A[i - 1] + 2 * K);
            ret = Math.min(ret, max - Math.min(A[i], A[0] + 2 * K));
        }
        return ret;
    }
}
// @lc code=end

